File: graph/union_find.py
class UF:
    def __init__(self, n):
        self.par = list(range(n))
        self.rank = [1] * n

    def find(self, n1):
        p1 = n1
        while p1 != self.par[p1]:
            self.par[p1] = self.par[self.par[p1]]
            p1 = self.par[p1]
        return p1

    def union(self, n1, n2):
        p1, p2 = self.find(n1), self.find(n2)
        if p1 == p2:
            return 0
        if self.rank[p2] > self.rank[p1]:
            p1, p2 = p2, p1
        self.par[p2] = p1
        self.rank[p1] += self.rank[p2]
        return 1

File: JavaImpl/src/main/java/org/example/util/UnionFind.java
package org.example.util;

class UnionFind {
  private final int[] parent;
  private final int[] rank;

  public UnionFind(int n) {
    parent = new int[n];
    rank = new int[n];
    for (int i = 0; i < n; i++) {
      parent[i] = i;
      rank[i] = 1;
    }
  }

  public int find(int n1) {
    if (n1 != parent[n1]) {
      parent[n1] = find(parent[n1]); // Path compression
    }
    return parent[n1];
  }

  public int union(int n1, int n2) {
    int p1 = find(n1);
    int p2 = find(n2);
    if (p1 == p2) {
      return 0;
    }
    if (rank[p2] > rank[p1]) {
      int temp = p1;
      p1 = p2;
      p2 = temp;
    }
    parent[p2] = p1;
    rank[p1] += rank[p2];
    return 1;
  }
}
